### PR TITLE
Card Parts Bottom Sheet View Controller

### PR DESCRIPTION
**Pull request by rcole34**
## Issue Link :link:
 <ul>
   <li> <i>Is this a bug fix or a feature?</i> Feature!</li>
   <li> <i>Does it break any existing functionality?</i> Nope.</li>
</ul>

## Goals of this PR :tada:
 <ul>
   <li> <i>Why is the change important?</i> Adds a new component that provides a new functionality to CardParts with extensive use cases.</li>
   <li> <i>How far has it been tested?</i> Extensive manual testing with some added unit tests.</li>
 </ul>
 
## How Has This Been Tested :mag:

Tested manually through examples in CardParts example app, and tested through unit tests.

## Test Configuration :space_invader:

<ul>
 <li> <i>Xcode version:</i> 11.2.1</li>
 <li> <i>Device/Simulator</i> iPhone11 Pro</li>
 <li> <i>iOS version</i> 13.2.2|| <i>MacOSX version</i> 10.14.6</li>
</ul>

## Things to check on :dart:

 - [x] My Pull Request code follows the coding standards and styles of the project 
 - [x] I have worked on unit tests and reviewed my code to the best of my ability 
 - [x] I have used comments to make other coders understand my code better 
 - [x] My changes are good to go without any warnings 
 - [x] I have added unit tests both for the happy and sad path 
 - [x] All of my unit tests pass successfully before pushing the PR 
 - [x] I have made sure all dependent downstream changes impacted by my PR are working

### CODE REVIEW

File: CardParts/src/Classes/CardPartsBottomSheetViewController.swift [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+    // MARK: Public functions
+    /// Adds the bottom sheet to the specified view, and makes calls to set up and display bottom sheet
+    /// - Parameter view: View on which to add the bottom sheet. Defaults to key window.
+    public func presentBottomSheet(on hostView: UIView? = UIApplication.shared.keyWindow) {
+        guard let hostView = hostView, !isShowingBottomSheet else { return }
+        self.view.translatesAutoresizingMaskIntoConstraints = false
+        
+        hostView.addSubview(self.view)
+        
+        NSLayoutConstraint.activate([
```
Comment (badrinathvm) [inline CardParts/src/Classes/CardPartsBottomSheetViewController.swift]:
```suggestion
        layout
```

File: CardParts/src/Classes/CardPartsBottomSheetViewController.swift [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+            self.view.leadingAnchor.constraint(equalTo: hostView.leadingAnchor),
+            self.view.trailingAnchor.constraint(equalTo: hostView.trailingAnchor),
+            self.view.bottomAnchor.constraint(equalTo: hostView.bottomAnchor)
+        ])
```
Comment (badrinathvm) [inline CardParts/src/Classes/CardPartsBottomSheetViewController.swift]: Try using layout from `LayoutProxy.swift`

File: CardParts/src/Classes/CardPartsBottomSheetViewController.swift [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+        calculateHeight()
+        
+        guard isShowingBottomSheet else { return }
+        // guard this because when reusing bottom sheet to avoid crash when sometimes constraints are mysteriously disappearing
+        guard self.bottomSheetTopConstraint != nil, self.viewTopConstraint != nil else {
+            self.dismissBottomSheet(.programmatic(info: ["error": "closing due to nil constraints"]))
+            return
+        }
+        UIView.animate(withDuration: changeHeightAnimationDuration, delay: 0, usingSpringWithDamping: 1.0, initialSpringVelocity: 1.0, options: UIView.AnimationOptions.curveEaseIn, animations: {
+            self.darkOverlay.alpha = self.overlayMaxAlpha
```
Comment (badrinathvm) [inline CardParts/src/Classes/CardPartsBottomSheetViewController.swift]: Can we allow user to animation type which they need ?

File: CardParts/src/Classes/CardPartsBottomSheetViewController.swift (hunk 2 above)
Comment (rcole34) [reply CardParts/src/Classes/CardPartsBottomSheetViewController.swift]: Good call, did this elsewhere but missed this one

File: CardParts/src/Classes/CardPartsBottomSheetViewController.swift (hunk 3 above)
Comment (rcole34) [reply CardParts/src/Classes/CardPartsBottomSheetViewController.swift]: Good idea, added a variable for this too 👍

File: CardParts/src/Classes/CardPartsBottomSheetViewController.swift [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+        viewTopConstraint = self.view.topAnchor.constraint(equalTo: hostView.topAnchor)
+        if !shouldIncludeOverlay {
+            viewTopConstraint = self.view.topAnchor.constraint(equalTo: hostView.bottomAnchor, constant: -bottomSheetHeight)
+        }
+        viewTopConstraint.isActive = true
+
+        self.view.layoutIfNeeded()
+        
+        // animate appearing on screen
+        UIView.animate(withDuration: appearAnimationDuration, delay: 0, usingSpringWithDamping: 1.0, initialSpringVelocity: 1.0, options: animationOptions, animations: {
```
Comment (chasecf) [inline CardParts/src/Classes/CardPartsBottomSheetViewController.swift]: Weakify self

File: CardParts/src/Extensions/UIView.swift [hunk 5]
```diff
@@ -6,6 +6,20 @@
 //
 
 extension UIView {
+    func roundedTop(radius: CGFloat){
```
Comment (chasecf) [inline CardParts/src/Extensions/UIView.swift]: Do we want it to just be `roundedTop` or a more generic `roundCorners` and they pass the corners they want to round?

File: Example/Tests/CardPartsBottomSheetTests.swift [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+//
+//  Created by Ryan Cole on 3/4/20.
+//  Copyright © 2020 CocoaPods. All rights reserved.
+//
+
+import XCTest
+
+@testable import CardParts
+
+class CardPartsBottomSheetTests: XCTestCase {
```
Comment (chasecf) [inline Example/Tests/CardPartsBottomSheetTests.swift]: Awesome!!!!

File: CardParts/src/Extensions/UIView.swift (hunk 5 above)
Comment (croossin) [inline CardParts/src/Extensions/UIView.swift]: Do we want to make this more generic instead of just top corners?

**Pull request merged**